### PR TITLE
refactor: All the warning set the warning type for future filter

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

ref #44551

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   -    |
| 🇨🇳 Chinese |    -     |

* ConfigProvider support `warning` config `deprecated=false` to ignore deprecated warning to help migration from v4.        * ConfigProvider 支持 `warning` 配置 `deprecated=false` 过滤废弃接口警告以帮助从 v4 迁移。  

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at eaaa394</samp>

This pull request improves the warning system for the ant-design components by adding a custom warning hook and context, using them in various components, and adding a warning prop and a warning configuration feature to the ConfigProvider component. It also updates some imports, types, props, and tests to support the new warning system and improve the code quality and consistency. The main files affected are `components/_util/warning.ts`, `components/config-provider/index.tsx`, and `components/config-provider/PropWarning.tsx`.

### <samp>🤖 Generated by Copilot at eaaa394</samp>

*  Add a custom warning hook and context for filtering warning messages based on the type and level of the warning ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-75853a37744088ef730adfb0906473685c80080827a097c12079581f39203ffbR1),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-75853a37744088ef730adfb0906473685c80080827a097c12079581f39203ffbR22-R54))
*  Replace the original warning function from rc-util with the devUseWarning hook in various components, such as Alert, Anchor, AutoComplete, Avatar, BackTop, and Breadcrumb, and pass the valid, component, type, and message parameters to the hook ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-18c21d0f3bf7f7e9f29dd782a9b585a896225e10fcc6d470f4de65e822de794cL121-R130),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L136-R147),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-584bab7d33a5231e7e6e844bdb79bf36b1b1abd13e03d1f45eaa3bf03d0f6ce7L54-R60),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL119-R121),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dR128),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dR135),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-a9110968bd15226a58d0270285b7d34c0c37c415a6e5ca231c84fb3fb82ef9efL135-R146),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-22587f2ed95b1c480472e426d68bdb6cc32d38b3134e9aa5a32101b973312618L54-R62),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f2426e3ced2642a4e369433c0c19b385fca436f57ad4620b4bd2bfcd7ab9254dL102-R134),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f2426e3ced2642a4e369433c0c19b385fca436f57ad4620b4bd2bfcd7ab9254dL173-R205))
*  Import React from 'react' in various files that use the devUseWarning hook or the WarningContext ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-18c21d0f3bf7f7e9f29dd782a9b585a896225e10fcc6d470f4de65e822de794cR1-R2),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-353746edad3f07acb175d7e32c64972b87359b5a0dc4e14e4d3f1133b61e48a8L1-R8),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-584bab7d33a5231e7e6e844bdb79bf36b1b1abd13e03d1f45eaa3bf03d0f6ce7L1-R4),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-a9110968bd15226a58d0270285b7d34c0c37c415a6e5ca231c84fb3fb82ef9efL1-R14),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f2426e3ced2642a4e369433c0c19b385fca436f57ad4620b4bd2bfcd7ab9254dL1-R16),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaL1-R6),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R3-R4))
*  Add the 'use client' directive to the files that rely on some browser-specific features or APIs, such as AutoComplete and BackTop ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL3-R12),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-22587f2ed95b1c480472e426d68bdb6cc32d38b3134e9aa5a32101b973312618L3-R13))
*  Add the warning prop to the ConfigProviderProps interface and the ConfigContextProps interface, and pass it to the memoedConfig object and the WarningContext.Provider component in `components/config-provider/index.tsx` ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaR159-R160),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R148-R150),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L334-R341),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R430),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R564-R570))
*  Add a PropWarning component that renders a warning message based on the props of the ConfigProvider component, and wrap the children prop with it in `components/config-provider/index.tsx` ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-77b3e18cb24de05844f60221d8618bf867dd6d7186f601b9abb28a10db361617R1-R30),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L42-R48),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L473-R476))
*  Add a test case for the warning support filter level feature of the ConfigProvider component, and import the resetWarned function from `../../_util/warning` to reset the internal state of the warning function or hook in `components/config-provider/__tests__/index.test.tsx` ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0240afe2faa6807b920ad4c86a45d6724eb47ed4cde840f097b69fdb4d28cdb5L1-R6),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0240afe2faa6807b920ad4c86a45d6724eb47ed4cde840f097b69fdb4d28cdb5R128-R136))
*  Add or remove some imports in various files to reflect the changes in the dependencies, such as importing devUseWarning, WarningContext, and WarningContextProps from `../_util/warning`, importing TabsProps and defaultTheme from `../tabs` and `../theme/context`, and removing useTheme and ShowWaveEffect ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-18c21d0f3bf7f7e9f29dd782a9b585a896225e10fcc6d470f4de65e822de794cL9-R14),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f7a3a5082a18fdd0627b75fb4f13fa559a2e5a5a781e381b8ba8c98b13318ccaL10-R16),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L8-R12),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56R22-R23),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L42-R48))
*  Remove some redundant warning function calls in various files, such as `components/auto-complete/index.tsx` and `components/breadcrumb/Breadcrumb.tsx`, since they are now handled by the devUseWarning hook or the PropWarning component ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dL107-L111),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-f2426e3ced2642a4e369433c0c19b385fca436f57ad4620b4bd2bfcd7ab9254dL173-R205),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-0a9895a32672ec55a084aa7165fc8666008942903b8e51e30b662e9e1ecffb56L334-R341))
*  Add some comments to the interface AutoCompleteProps in `components/auto-complete/index.tsx` to indicate that the dataSource prop is deprecated and suggest using the options prop instead ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dR39))
*  Add the type parameter to some devUseWarning hook calls in `components/auto-complete/index.tsx` to specify the type of the warning, such as 'usage' or 'deprecated' ([link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dR128),[link](https://github.com/ant-design/ant-design/pull/44613/files?diff=unified&w=0#diff-c55203058d3a4064d8f52a5b4e7d7770153d215966d276d6814011d99268f76dR135))

### CODE REVIEW

**Review by MadCcc — APPROVED** [review]
LGTM

File: components/auto-complete/index.tsx [hunk 1]
```diff
@@ -102,33 +104,33 @@ const AutoComplete: React.ForwardRefRenderFunction<RefSelectProps, AutoCompleteP
...
     );
 
     warning(
       !customizeInput || !('size' in props),
       'AutoComplete',
+      'usage',
       'You need to control style self instead of setting `size` when using customize input.',
     );
 
     warning(
```
Comment (kiner-tang) [inline components/auto-complete/index.tsx:130]: 感觉是不是可以基于这个方法在专门封装一个用来过期属性的警告方法，只需要穿旧属性和新属性名就好了，还可以保证过期提示文案的统一，如：
```ts
deprecatedWarning(
     !dropdownClassName,// 展示条件
      'AutoComplete',// 组件名
      'dropdownClassName', // 旧属性
      'popupClassName',// 新属性
    );
```

File: components/auto-complete/index.tsx [hunk 2]
```diff
@@ -102,33 +104,33 @@ const AutoComplete: React.ForwardRefRenderFunction<RefSelectProps, AutoCompleteP
...
-                '`dataSource` is only supports type `string[] | Object[]`.',
-              );
               return undefined;
           }
         })
       : [];
   }
 
   if (process.env.NODE_ENV !== 'production') {
+    const warning = devUseWarning();
```
Comment (kiner-tang) [inline components/auto-complete/index.tsx:114]: 是不是可以在这里传入组件名呢，后面就不需要传那么多了，节省一点大小：
```ts
const warning = devUseWarning('AutoComplete');
```

File: components/auto-complete/index.tsx (hunk 1 above)
Comment (zombieJ) [reply components/auto-complete/index.tsx:130]: 可以的，可以开个新的 PR 搞。这个 PR 现在只做全部的替换，不加其他逻辑

File: components/auto-complete/index.tsx (hunk 2 above)
Comment (zombieJ) [reply components/auto-complete/index.tsx:114]: 这个打包的时候会全部干掉，不影响 bundle size。不过这个点子不错，省手力

File: components/auto-complete/index.tsx (hunk 1 above)
Comment (kiner-tang) [reply components/auto-complete/index.tsx:130]: > 可以的，可以开个新的 PR 搞。这个 PR 现在只做全部的替换，不加其他逻辑

这种粗活我最积极了

**Pull request merged**